catalyst-gateway-crates/c509-certificate/src/c509_oid.rs

```diff
@@ -0,0 +1,213 @@
+//! C509 OID provides an encoding and decoding of C509 Object Identifier (OID).
+//! Please refer to [RFC9090](https://datatracker.ietf.org/doc/rfc9090/) for OID encoding
+//! Please refer to [CDDL Wrapping](https://datatracker.ietf.org/doc/html/rfc8610#section-3.7)
+//! for unwrapped types.
+
+use anyhow::Result;
+use asn1_rs::oid;
+use minicbor::{data::Tag, decode, encode::Write, Decode, Decoder, Encode, Encoder};
+use oid_registry::Oid;
+
+use crate::tables::IntegerToOidTable;
+
+/// IANA Private Enterprise Number (PEN) OID prefix.
+const PEN_PREFIX: Oid<'static> = oid!(1.3.6 .1 .4 .1);
+
+/// Tag number representing IANA Private Enterprise Number (PEN) OID.
+const OID_PEN_TAG: u64 = 112;
+
+/// A strut of C509 OID with Registered Integer Encoding/Decoding.
+#[derive(Debug, Clone, PartialEq)]
+pub struct C509oidRegistered {
+    /// The `C509oid`.
+    oid: C509oid,
+    /// The registration table.
+    registration_table: &'static IntegerToOidTable,
+}
+
+impl C509oidRegistered {
+    /// Create a new instance of `C509oidRegistered`.
+    pub(crate) fn new(oid: Oid<'static>, table: &'static IntegerToOidTable) -> Self {
+        C509oidRegistered {
+            oid: C509oid::new(oid),
+            registration_table: table,
+        }
+    }
+
+    /// Is PEN Encoding supported for this OID.
+    /// Depends on each registration table.
+    pub(crate) fn pen_encoded(mut self) -> Self {
+        self.oid.pen_supported = true;
+        self
+    }
+
+    /// Get the `C509oid`.
+    pub(crate) fn get_c509_oid(&self) -> C509oid {
+        self.oid.clone()
+    }
+
+    /// Get the registration table.
+    pub(crate) fn get_table(&self) -> &'static IntegerToOidTable {
+        self.registration_table
+    }
+}
+
+// -----------------------------------------
+
+/// A struct represent an instance of `C509oid`.
+#[derive(Debug, PartialEq, Clone, Eq, Hash)]
+pub struct C509oid {
+    /// The OID.
+    oid: Oid<'static>,
+    /// The flag to indicate whether PEN encoding is supported.
+    pen_supported: bool,
+}
+
+impl C509oid {
+    /// Create an new instance of `C509oid`.
+    /// Default value of PEN flag is false
+    #[must_use]
+    pub fn new(oid: Oid<'static>) -> Self {
+        C509oid {
+            oid,
+            pen_supported: false,
+        }
+    }
+
+    /// Is PEN Encoding supported for this OID
+    #[must_use]
+    pub fn pen_encoded(mut self) -> Self {
```

Comment:
After updating `C509oid::new` remove this

Reply:
ill change to `pub (crate)`